Mathlib/CategoryTheory/Distributive/Monoidal.lean

```diff
@@ -0,0 +1,321 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sina Hazratpour
+-/
+
+
+import Mathlib.CategoryTheory.Closed.Monoidal
+import Mathlib.CategoryTheory.Limits.Preserves.Shapes.BinaryProducts
+import Mathlib.CategoryTheory.Limits.FunctorCategory.Shapes.BinaryProducts
+import Mathlib.CategoryTheory.Monoidal.Braided.Basic
+import Mathlib.CategoryTheory.Monoidal.End
+
+/-!
+
+# Distributive monoidal categories
+
+## Main definitions
+
+A monoidal category `C` with binary coproducts is left distributive if the left tensor product
+preserves binary coproducts. This means that, for all objects `X`, `Y`, and `Z` in `C`,
+the cogap map `(X ⊗ Y) ⨿ (X ⊗ Z) ⟶ X ⊗ (Y ⨿ Z)` can be promoted to an isomorphism. We refer to
+this isomorphism as the left distributivity isomorphism.
+
+A monoidal category `C` with binary coproducts is right distributive if the right tensor product
+preserves binary coproducts. This means that, for all objects `X`, `Y`, and `Z` in `C`,
+the cogap map `(Y ⊗ X) ⨿ (Z ⊗ X) ⟶ (Y ⨿ Z) ⊗ X` can be promoted to an isomorphism. We refer to
+this isomorphism as the right distributivity isomorphism.
+
+A distributive monoidal category is a monoidal category that is both left and right distributive.
+
+## Main results
+
+- A symmetric monoidal category is left distributive if and only if it is right distributive.
+
+- A closed monoidal category is left distributive.
+
+- For a category `C` the category of endofunctors `C ⥤ C` is left distributive (but almost
+  never right distributive). The left distributivity is tentamount to the fact that the coproduct
+  in the functor categories is computed pointwise.
+
+## TODO
+
+Show that a distributive monoidal category whose unit is weakly terminal is finitary distributive.
+
+Provide more examples of the distributive monoidal structure on the following categories:
+
+- The category of abelian groups with the monoidal structure given by the tensor product of
+  abelian groups.
+- The category of R-modules with the monoidal structure given by the tensor product of modules.
+- The category of vector bundles over a topological space where the monoidal structure is given by
+  the tensor product of vector bundles.
+- The category of pointed types with the monoidal structure given by the smash product of
+  pointed types and the coproduct given by the wedge sum.
+
+## References
+
+[Hans-Joachim Baues, Mamuka Jibladze, Andy Tonks, Cohomology of
+ monoids in monoidal categories, in: Operads: Proceedings of Renaissance
+ Conferences, Contemporary Mathematics 202, AMS (1997) 137-166][]
+
+-/
+
+universe v v₂ u u₂
+
+noncomputable section
+
+namespace CategoryTheory
+
+open Category MonoidalCategory Limits Iso
+
+/-- A monoidal category with binary coproducts is left distributive
+if the left tensor product functor preserves binary coproducts. -/
+class IsMonoidalLeftDistrib (C : Type u) [Category.{v} C]
+    [MonoidalCategory.{v} C] [HasBinaryCoproducts C] : Prop where
```

Comment:
The universe parameter of `MonoidalCategory` can be inferred from the category structure:
```suggestion
    [MonoidalCategory C] [HasBinaryCoproducts C] : Prop where
```